### ISSUE TITLE
fonttools memory usage much higher than expected

### ISSUE THREAD

**Issue by cjchapman** [fonttools/fonttools#1515]
@graphicore, @webdev, and I were looking into Font Bakery's memory usage and found that fonttools' memory usage was much higher than we expected, and @anthrotype encouraged us to report this here.

The discussion started in the Font Bakery Gitter on 12 Feb. 2019 and continued through at least 22 Feb 2019: [February 12, 2019 11:22 AM](https://gitter.im/fontbakery/Lobby?at=5c631ced253c2b5ea3ff63a0)

Here's the test script that Lasse wrote:
https://gist.github.com/graphicore/7618637aefbc474f6d69a294dab86616

Here's the memory usage graph for `NotoSansMongolian-Regular.ttf` -- it's a 136 KB font for which fonttools is using about 40 MB of memory:
![alt text](https://files.gitter.im/fontbakery/Lobby/zeRx/Screenshot-from-2019-02-21-19-55-45.png "Noto Sans Mongolian")
The command lines to reproduce the above image are:
```sh
mprof run ttfont-decompile.py NotoSansMongolian-Regular.ttf
mprof plot
```

Here's the memory usage graph for `SourceHanSans-Regular.otf` -- it's a 16 MB font for which fonttools is using over 120 MB of memory:
![alt text](https://files.gitter.im/fontbakery/Lobby/xjDG/Figure-1_226.png "Source Han Sans")
The command lines to reproduce the above image are:
```sh
mprof run ttfont-decompile.py SourceHanSans-Regular.otf
mprof plot
```

**Comment by justvanrossum** [comment]
That's odd, I get much lower numbers for the first test (macOS 10.10):

![image](https://user-images.githubusercontent.com/4246121/53262765-cdcacf80-36d7-11e9-92a6-c824d39b4c09.png)

**Comment by justvanrossum** [comment]
But only slightly lower for SourceHanSans:
![image](https://user-images.githubusercontent.com/4246121/53262871-12ef0180-36d8-11e9-975a-4f9c6d36ded5.png)

**Comment by cjchapman** [comment]
@justvanrossum I also got lower numbers on my Mac than Lasse did on his Linux box. The images I posted are from Lasse's Linux box.

**Comment by behdad** [comment]
I would say this is expected.

The first hump seems to be 1. python startup + reading the font file into memory (which the TTFont constructor does unnecessarily; there's an issue for that).  On my Linux box, running mprof on an empty script shows 32mb memory usage, which is consistent with Lasse's graphs.

The second hump, is loading all tables:

  - For 'glyf' table, that's reading all glyf binaries but not expanding them; so that's not too bad.
  
  - For 'CFF' table charstrings, every one or two bytes turns into a string,

  - For GSUB/GPOS tables, almost every two bytes of the table turns into an object.

So I don't see how we can optimize the memory usage further, other than the existing "lazy=True" option to TTFont constructor.

I'm leaning towards closing this as Not Planned.

**Comment by Lucas-C** [comment]
> reading the font file into memory (which the TTFont constructor does unnecessarily; there's an issue for that)

I think that is the relevant issue: https://github.com/fonttools/fonttools/issues/2252

**Comment by anthrotype** [comment]
@Lucas-C thanks for the link. Are you encountering this problem too?
Have you tried opening the TTFont with lazy=True like Behdad suggested in the previous comment?

**Comment by Lucas-C** [comment]
Yes 😊
I was actually able to reduce the memory usage of `fpdf2` by 10MiB by using `lazy=True`.
Relevant PR: https://github.com/PyFPDF/fpdf2/pull/718

I don't think that is the root cause of the memory leak I'm chasing though (_cf._ https://github.com/PyFPDF/fpdf2/pull/718), but that's still a win!

`fonttools` is really great, but I wonder: why isn't `lazy=True` the default value?

**Comment by anthrotype** [comment]
> why isn't lazy=True the default value?

I'd say historical, was probably a feature added later on and it's too late to change a default value like that. We did find sometimes find bugs that'd only show up when lazy=True, fixed now, but still I don't want risk suddenly break thing by change the default.

I'm curious, what font is showing the excessive memory usage in your testing?

**Comment by Lucas-C** [comment]
> I'm curious, what font is showing the excessive memory usage in your testing?

The memory usage was especially high with `NotoColorEmoji.ttf`:
https://github.com/PyFPDF/fpdf2/issues/641#issuecomment-1463639032

**Comment by anthrotype** [comment]
well.. that's a huge CBDT/CBLC color font containing thousands of PNGs
are you rendering them all of them?

**Comment by anthrotype** [comment]
ok looks like [CBDT table](https://github.com/fonttools/fonttools/blob/48bf4d4eede76209a486579db04f7091863860c5/Lib/fontTools/ttLib/tables/E_B_D_T_.py#L51) (which inherits from EBDT) doesn't implement lazy glyph loading, but decompiles them all at once... /cc @behdad

**Comment by Lucas-C** [comment]
> well.. that's a huge CBDT/CBLC color font containing thousands of PNGs are you rendering them all of them?

No, I'm not rendering all of them, but that does not have much to do with the problem I faced.

As can be seen in the sample code I pointed to, the memory allocated by `fonttools` when parsing this font file is NOT garbage collected properly (if `lazy=True` is not provided).
